bigquery/google/cloud/bigquery/table.py

```diff
@@ -1402,17 +1443,46 @@ def _to_dataframe_bqstorage(self, bqstorage_client, dtypes):
         if not session.streams:
             return pandas.DataFrame(columns=columns)
 
-        def get_dataframe(stream):
-            position = bigquery_storage_v1beta1.types.StreamPosition(stream=stream)
-            rowstream = bqstorage_client.read_rows(position)
-            return rowstream.to_dataframe(session, dtypes=dtypes)
+        # Use _to_dataframe_finished to notify worker threads when to quit.
+        # See: https://stackoverflow.com/a/29237343/101923
+        self._to_dataframe_finished = False
```

Comment:
Mostly I'm trying to reason about scope here.  This is a worker pool, but is it possible we're generating multiple independent dataframes that would share the same access to _to_dataframe_finished?  Do we need to key the workers to specific invocations, or is the nature of the access always blocking so that this isn't an issue?

Reply:
There are a couple reasons that I don't think it's an issue:

* Yes, `to_dataframe()` is a blocking call, so you wouldn't really have multiple going at once.
* `RowIterator` isn't really something you'd want to use across threads or even more than once anyway. Because of how the pagination state works, once you loop through all the rows once, `to_dataframe()` returns an empty DataFrame, even in the case where BQ Storage API isn't used.